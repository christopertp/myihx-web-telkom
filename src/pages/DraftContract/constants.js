export const FAILED = 'DraftContract/failed';
export const FETCHED = 'DraftContract/fetched';
export const LOADING = 'DraftContract/loading';
