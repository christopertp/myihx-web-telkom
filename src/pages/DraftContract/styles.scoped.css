.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 1rem;
}

.root {
  background-color: #ffffff;
  border-radius: 0.5rem;
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  margin: 0 auto;
  padding: 1.5rem;
  width: 50rem;
}

.root > header > h1 {
  color: #333435;
  font-size: 1rem;
  letter-spacing: 0.0625rem;
  margin-bottom: 2rem;
  text-align: center;
  text-transform: uppercase;
}

.spinner {
  display: block;
  margin: auto;
}

.root > footer {
  display: flex;
  margin-top: 2rem;
}

.root > footer > * { flex: 1; }
.root > footer > *:first-child { margin-right: 1rem; }

@media only screen and (max-width: 48em) {
  .pdf {
    flex: 1;
  }

  .root {
    display: flex;
    flex-direction: column;
    height: 100vh;
    margin: -1rem;
    padding: 0;
    width: 100vw;
  }

  .root > header {
    align-items: center;
    display: flex;
    padding: 1.25rem 1rem;
    width: 100vw;
  }

  .root > header > h1 {
    margin: 0 auto;
  }

  .root > footer {
    margin: 0;
    padding: 1rem;
  }  
}
