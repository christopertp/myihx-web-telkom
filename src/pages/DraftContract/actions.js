import { push } from 'react-router-redux';
import { draftContract, draftContractAccept } from '../../services/user';
import { FAILED, FETCHED, LOADING } from './constants';

export function fetchDraftContract() {
  return async dispatch => {
    dispatch(loadingAction(true, 'pdf'));
    try {
      const { data } = await draftContract();
      dispatch(fetchedAction(data.draftContract));
    } catch (err) {
      dispatch(failedAction('Gagal memuat PDF', 'pdf'));
    }
  };
}

export function fetchAccept() {
  return async dispatch => {
    dispatch(loadingAction(true, 'submit'));
    try {
      await draftContractAccept();
      dispatch(failedAction('', 'submit'));
      dispatch(push({ pathname: '/', state: { notif: 'Draft contract is accepted.' } }));
    } catch (err) {
      dispatch(failedAction(err.message, 'submit'));
    }
  };
}

function failedAction(message, key) {
  return { type: FAILED, key,  message };
}

function fetchedAction(url) {
  return { type: FETCHED, url };
}

function loadingAction(isLoading, key) {
  return { type: LOADING, isLoading, key };
}
