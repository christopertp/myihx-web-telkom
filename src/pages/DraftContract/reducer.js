import { FAILED, FETCHED, LOADING } from './constants';

const initialState = {
  isLoading: {
    p: true,
    s: false,
  },
  message: '',
  url: '',
};

export default function reducer(state = initialState, action = {}) {
  const { type, isLoading, key, message, url } = action;

  switch (type) {
    case FAILED:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key.charAt(0)]: false },
        message,
      };
    case FETCHED:
      return {
        ...state,
        isLoading: { ...state.isLoading, p: false },
        message: '',
        url,
      };
    case LOADING:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key.charAt(0)]: isLoading },
      };
    default:
      return state;
  }
}
