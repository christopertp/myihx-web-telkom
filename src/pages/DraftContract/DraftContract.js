import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Button from '../../components/elements/Button';
import IconButton from '../../components/elements/IconButton';
import PdfViewer from '../../components/elements/PdfViewer';
import Spinner from '../../components/elements/Spinner';
import { fetchAccept, fetchDraftContract } from './actions';
import styles from './styles.scoped.css';

export default function DraftContract() {
  const dispatch = useDispatch();
  const { isLoading, message, url } = useSelector(s => s.draftContract);

  useEffect(() => {
    dispatch(fetchDraftContract());
  }, []);

  return (
    <section className={styles.root}>
      <header>
        <IconButton name="back_red" to="/" />
        <h1>Review Draft Contract</h1>
      </header>
      {(message !== 'Gagal memuat PDF') && <PdfViewer className={styles.pdf} url={url} />}
      {message && <small className={styles.error}>{message}</small>}
      <footer>
        <Button to="/" variant="bordered">Decline</Button>
        <Button disabled={!url || isLoading.s} onClick={() => dispatch(fetchAccept())}>
          {isLoading.s ? <Spinner /> : 'Accept'}
        </Button>
      </footer>
    </section>
  );
}
