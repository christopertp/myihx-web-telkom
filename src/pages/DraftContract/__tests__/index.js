import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../components/elements/Spinner';
import DraftContract from '../DraftContract';

jest.mock('../actions', () => ({
  fetchAccept: jest.fn(),
  fetchDraftContract: jest.fn(),
}));

useSelector.mockImplementation(fn => {
  fn({});
  return {
    isLoading: { s: false },
    message: '',
    url: 'url',
  };
});

describe('src/pages/DraftContract', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<DraftContract />);
    expect(tree).toMatchSnapshot();
  });

  test('render loading state, error state, & onClick', () => {
    const isLoading = { s: true };
    const message = 'error';

    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    useSelector.mockImplementationOnce(() => ({ isLoading, message }));
    const result = DraftContract();

    expect(result.props.children[2].props.children).toBe('error');

    const btnAccept = result.props.children[3].props.children[1];
    btnAccept.props.onClick();

    expect(btnAccept.props.children.type).toBe(Spinner);
    expect(dispatch).toHaveBeenCalled();
  });
});
