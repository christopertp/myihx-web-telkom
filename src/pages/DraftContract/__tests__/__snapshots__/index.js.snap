// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/DraftContract render 1`] = `
<section>
  <header>
    <IconButton
      className=""
      name="back_red"
      onClick={[Function]}
      title=""
      to="/"
      type="button"
    />
    <h1>
      Review Draft Contract
    </h1>
  </header>
  <PdfViewer
    className=""
    url="url"
  />
  
  <footer>
    <Button
      className=""
      disabled={false}
      onClick={null}
      to="/"
      type="button"
      variant="bordered"
    >
      Decline
    </Button>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      Accept
    </Button>
  </footer>
</section>
`;
